### ISSUE TITLE
flycheck-eslint errors out with a bare setup

### ISSUE THREAD

**Issue by cowboyd** [flycheck/flycheck#1085]
Opened up a file with a .js extension using the javascript-eslint checker enabled. It plopped a big nasty error into my minibuffer saying that the checker might be mal-defined instead of verifying the javascript syntax as expected.

<img width="1222" alt="emacs_prelude_-_scratch_js_and_check_buffers_ _flycheck_30-cvs_documentation_" src="https://cloud.githubusercontent.com/assets/4205/18453585/93de9c44-7905-11e6-9ee2-ccf65eda6fc1.png">

It looks as though the problem is that `eslint` throws an error if you don't have at least some configuration file present, even if it's an empty one in your home directory. This seems suboptimal to me, but I don't know if this is something that I should raise upstream, or that can be worked around in flycheck.

Elsint does support a `--no-eslintrc` flag that will make the error go away:

```
$ echo "var x = 'y';" | eslint --stdin                                                                                                                                        
No ESLint configuration found.
Error: No ESLint configuration found.
    at getLocalConfig (/Users/cowboyd/.nvm/versions/node/v4.2.3/lib/node_modules/eslint/lib/config.js:162:35)
    at Config.getConfig (/Users/cowboyd/.nvm/versions/node/v4.2.3/lib/node_modules/eslint/lib/config.js:256:22)
    at processText (/Users/cowboyd/.nvm/versions/node/v4.2.3/lib/node_modules/eslint/lib/cli-engine.js:222:33)
    at CLIEngine.executeOnText (/Users/cowboyd/.nvm/versions/node/v4.2.3/lib/node_modules/eslint/lib/cli-engine.js:749:26)
    at Object.execute (/Users/cowboyd/.nvm/versions/node/v4.2.3/lib/node_modules/eslint/lib/cli.js:174:42)
    at /Users/cowboyd/.nvm/versions/node/v4.2.3/lib/node_modules/eslint/bin/eslint.js:57:36
    at ConcatStream.<anonymous> (/Users/cowboyd/.nvm/versions/node/v4.2.3/lib/node_modules/eslint/node_modules/concat-stream/index.js:36:43)
    at emitNone (events.js:72:20)
    at ConcatStream.emit (events.js:166:7)
    at finishMaybe (/Users/cowboyd/.nvm/versions/node/v4.2.3/lib/node_modules/eslint/node_modules/concat-stream/node_modules/readable-stream/lib/_stream_writable.js:475:14)
```

vs

```
$ echo "var x = 'y';" | eslint --stdin --no-eslintrc
```

Perhaps there is a way to detect this and retry with the `--no-eslintrc` set?

## Environment

```
Syntax checkers for buffer scratch.js in js2-mode:

  javascript-eslint
    - predicate:          t
    - executable:         Found at /Users/cowboyd/.nvm/versions/node/v4.2.3/bin/eslint
    - configuration file: Not found

  javascript-jshint
    - predicate:          nil
    - executable:         Not found
    - configuration file: Not found

  javascript-gjslint
    - predicate:          nil
    - executable:         Not found
    - configuration file: Not found

  javascript-jscs
    - predicate:          nil
    - executable:         Not found
    - configuration file: Not found

  javascript-standard
    - predicate:  nil
    - executable: Not found

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 30snapshot (package: 20160912.814)
Emacs version:    24.4.1
System:           x86_64-apple-darwin14.0.0
Window system:    ns
```

```
GNU Emacs 24.4.1 (x86_64-apple-darwin14.0.0, NS apple-appkit-1343.14) of 2014-10-27 on galadriel.att.net
```

```
Flycheck version: 30snapshot (package: 20160912.814)
```

**Comment by lunaryorn** [comment]
Oh I wasn't aware that eslint now errors if it doesn't find a configuration.  Unfortunately there's little that we can do:  We have no infrastructure to re-run a syntax checker with different flags if a particular error occurs, and I don't want to add a special case for just a single syntax checker.

Defaulting to `--no-eslintrc` is not an option either as it'd completely disable configuration files even if they exist as far as I understand.  We could only add a user option for it.

The best way though would be to take this issue to eslint and ask them to gracefully deal with the absence of a configuration file.  I'd expect eslint to do at least a syntax check of some kind if there's no configuration file.

In any case we should probably turn our default back to jshint.

@flycheck/javascript  Your opinion?

**Comment by aaronjensen** [comment]
Can we add a search for an eslintrc to the predicate?

**Comment by lunaryorn** [comment]
@aaronjensen I'm not sure.  eslint is looking at a lot of places for its configuration—including package.json—and I'm not sure in how far we'd have to deal with its [intricate cascading of config files](http://eslint.org/docs/user-guide/configuring#configuration-cascading-and-hierarchy).

**Comment by Simplify** [comment]
Ok, this is change in one of the recent versions of ESLint. This problem does not occur with 2.8 that I had installed, but after installing 3.5 I got same error. As @lunaryorn said, we can not use predicate, there are different ways to put ESLint configuration. So there is also no way to "auto magically" add `--no-eslintrc`. We can add option for this flag, but that is really project dependent and users will need to use that option in `.dir-locals.el` on project basis.

Strange thing is that according to ESLint [rules page](http://eslint.org/docs/rules/): "No rules are enabled by default.".  If there is "default" that are no rules enabled, why `--no-eslintrc` option exists? Just to return '0' back to shell? Does not compute. That page is or outdated or this is a bug in ESLint. Also, if somebody specify JSON output as option to ESLint, that error has to be returned in JSON, simply because who ever called ESLint with that option expect JSON back and will not be able to parse response. This is bug IMHO.

We can:
* Change default checker to JSHint. I have no problem with that, personally, JSHint is default Ember.js checker, Ember's test suite assert that all files are validated with JSHint with .jshintrc files generated with Ember project. However, ESLint is currently the best JS checker available and used by most React projects (and more). This is really last valid option.
* Build predicate and require users to use `.eslintrc` and ignore 5 other ways to configure ESLint. I don't like this one. This will break many existing configurations.
* Add in documentation that user must generate config file with `eslint --init` in project or in his home directory before using ESLint. That will not stop crash issue...
* Add option for `--no-eslintrc` but I think that this is just waste of time and brings more complexity.
* The best option is as @lunaryorn said, to take this issue upstream, to ESLint developers.

**Comment by lunaryorn** [comment]
@Simplify Yup, I'd love to have eslint people weight in on this issue.

**Comment by Simplify** [comment]
Ok, I'll open issue there today. Can't do it right now doe some other obligations.

**Comment by cowboyd** [comment]
> Change default checker to JSHint. I have no problem with that, personally, JSHint is default Ember.js checker, Ember's test suite assert that all files are validated with JSHint with .jshintrc files generated with Ember project. However, ESLint is currently the best JS checker available and used by most React projects (and more). This is really last valid option.

While it's still the default, eslint support for Ember is really good, and I believe there are plans to make eslint the default for new projects soon. In fact, I actually enountered this because of using eslint within Ember projects.

**Comment by Simplify** [comment]
@cowboyd No problem at all, even I already started creating issue there :) I'll add my 5 cents there.

**Comment by Simplify** [comment]
eslint/eslint#7144

**Comment by Simplify** [comment]
@cowboyd The best option for now is to create config in your project and place it in version control. That way everybody on your team will have same config and not have this issue.

**Comment by lunaryorn** [comment]
To summarize, we'll

1. add #1089 (hence `blocked`),
2. add an `:enabled` function for eslint that uses `eslint --print-config` to check whether the current buffer has a suitable configuration,
3. and thus implicitly fall back to jshint or standard if there's none.

/cc @Simplify @flycheck/maintainers

**Comment by eastwood** [comment]
I believe this issue has caused eslint to be extremely slow with `--print-config`. Opening any buffer now has significant latency, almost unusable :(

**Comment by fmdkdd** [comment]
@eastwood See #1129 and its discussion.

**Comment by tko** [comment]
It seems this broke compatibility with older eslint versions that do not support `--print-config` option. With .eslintrc in place in the project everything used to work fine but no more.

```sh
$ eslint --print-config .
Invalid option '--print-config' - perhaps you meant '--inline-config'?
$ eslint --version
v1.10.3
```

I realize 1.10.3 is relatively old (Dec 1, 2015) so not sure if there's enough interest to keep compatibility.

**Comment by Simplify** [comment]
@tko we don't support versions that old. You need at `eslint` version that supports `--print-config` (3.0 I think, maybe even 3.5).
